lib/SILOptimizer/Mandatory/TFDevicePartition.cpp

```diff
@@ -255,14 +256,49 @@ class DevicePartitionCloner
 };
 }  // end anonymous namespace
 
+void DevicePartitionCloner::visitGraphOperationInst(GraphOperationInst *inst) {
+  if (!targetOps.count(inst)) return;
+
+  // TODO: visitTensorTransferInst?
+
+  GraphOperationDecoder decoder(inst);
+  auto deviceType = decoder.getDeviceType();
+
+  // Skip this instruction if it isn't for the current device.
+  if (deviceType != DeviceType::ALL && deviceType != thisDeviceType)
+    return;
+
+  auto &B = getBuilder();
+  auto loc = remapLocation(getUserSourceLocation(inst->getDebugLocation()));
+
+  // Clone it over.
+  SmallVector<SILValue, 4> args;
+  for (unsigned i = 0, e = inst->getNumOperands(); i != e; ++i) {
+    auto opValue = inst->getOperand(i);
+    assert(isTensorFlowValue(opValue->getType()) &&
+           "ops should only use tensors");
+    args.push_back(remapValue(opValue));
+  }
+
+  SmallVector<SILType, 2> resultTypes;
+  for (auto r : inst->getResults())
+    resultTypes.push_back(r->getType());
+
+  auto newOp = B.createGraphOperation(loc, inst->getName(), args,
+                                      inst->getAttributes(), resultTypes);
```

Comment:
the use of "inst->getAttributes()" has an awesome pay-off in simplicity, compared to the tfop builtin!

Reply:
Yes, it's a huge improvement!